e2e/cluster/cluster.go

```diff
@@ -0,0 +1,503 @@
+package cluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Azure/agentbakere2e/config"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/containerservice/armcontainerservice"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
+)
+
+// WARNING: if you modify cluster configuration, please change the version below
+// this will avoid potential conflicts with tests running on other branches
+// there is no strict rules or a hidden meaning for the version
+// testClusterNamePrefix is also used for versioning cluster configurations
+const testClusterNamePrefix = "abe2e-v20240704-"
+
+var (
+	clusterKubenet       *Cluster
+	clusterKubenetAirgap *Cluster
+	clusterAzureNetwork  *Cluster
+
+	clusterKubenetError       error
+	clusterKubenetAirgapError error
+	clusterAzureNetworkError  error
+
+	clusterKubenetOnce       sync.Once
```

Comment:
could we just use a singleflight group instead of managing a bunch of different `Once`'s?

Reply:
eh, maybe not, singleflight groups don't guarantee exactly-once execution, it only suppresses concurrent duplicate calls

Reply:
actually, are the `Once`'s here necessary at all? don't we create missing test clusters before running any scenarios?

Reply:
I don't mind a small number of duplicates. Especially when they located close to each other.
I can avoid id with something similar to this https://github.com/Azure/AgentBaker/blob/master/e2e/config/vhd.go#L83-L100

But it will introduce additional complexity and makes it hard to understand. 3 clusters aren't much to maintain.

Reply:
> actually, are the Once's here necessary at all? don't we create missing test clusters before running any scenarios?

This logic has changed a bit.

I have another long-term goal. I'd like each test to be independently executable from the others. I want to try organizing tests differently, with better integration of dev tooling.

Currently, there is a large, shared preparation step, and essentially a cluster setup for one test can break other tests that don't require such a cluster.